### PR TITLE
ARROW-8167: [CI] Add support for skipping builds with skip pattern in pull request title

### PR DESCRIPTION
**Pull request by kszucs**
In certain cases, especially when someone adds a new build configuration it can be handy to disable all but that job.

Sadly the commit message is not available for the pull_request event, so we cannot have conditions on it, but we can use the pull request's title for this purpose.

Hopefully the pull request event payload will contain the head commit in the future.

### CODE REVIEW

**Comment by kszucs** [comment]
Hmm, it is pretty annoying. The commit message is not available for the pull_request event, so we cannot have conditions on it.

File: .github/workflows/archery.yml [hunk 1]
```diff
@@ -30,6 +30,7 @@ on:
 jobs:
 
   archery:
+    if: ${{ !contains(github.event.pull_request.title, '[skip ci]') }}
```
Comment (kou) [inline .github/workflows/archery.yml]: I think that  we don't need `${{ }}` in `if`.

**Review by kou — COMMENTED** [review]
I'll consider how to implement this by commit message instead of pull request title.
Please wait a while.

File: .github/workflows/archery.yml (hunk 1 above)
Comment (kszucs) [reply .github/workflows/archery.yml]: @kou yeah, according to the GHA documentation we indeed don't need the curly braces, but nothing has been triggered until I added `${{ }}` around the expression...

**Comment by kszucs** [comment]
> I'll consider how to implement this by commit message instead of pull request title.
> Please wait a while.

My findings are:
- The push event gives the commit message, but this event only happens when someone directly pushes to a branch. We always want to run CI on master or any other `maint-*` branch, so we shouldn't skip any builds on push event.
- The pull request event is triggered when a commit arrives to an active pull request, but the pull request event's payload doesn't contain the commit's message, only the sha.

We can use the sha in two ways:
1. Create a script which queries the API and halts the build it the skip pattern is present. 
   **Drawback:** the builds would be still triggered whereas using the `if: ` expression for a job doesn't even try to run the build if the expression evaluates to false.
2. Create a build where we solely check that condition using the same script and make the rest of the jobs dependent on it with the [needs options](https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idneeds).
  **Drawback**: All of our builds would be triggered with a delay equal to the build time of the skip pattern checker, which means even more builds and slower triggering.

For developing new builds temporarily adding the skip pattern to the pull requests title seems like a good compromise to me.

File: .github/workflows/archery.yml (hunk 1 above)
Comment (kou) [reply .github/workflows/archery.yml]: Oh...

File: .github/workflows/cpp_cron.yml [hunk 2]
```diff
@@ -144,7 +147,7 @@ jobs:
   ubuntu-cmake32:
     name: AMD64 Ubuntu 18.04 C++ CMake 3.2
     runs-on: ubuntu-latest
-    if: github.event_name == 'push'
+    if: ${{ !contains(github.event.pull_request.title, '[skip ci]') }}
```
Comment (kou) [inline .github/workflows/cpp_cron.yml]: Do we need this change?

File: .github/workflows/csharp.yml [hunk 3]
```diff
@@ -86,7 +88,7 @@ jobs:
   macos:
     name: AMD64 MacOS 10.15 C# ${{ matrix.dotnet }}
     runs-on: macos-latest
-    if: github.event_name == 'push'
+    if: !contains(github.event.pull_request.title, '[skip ci]')
```
Comment (kou) [inline .github/workflows/csharp.yml]: Do we need this change?

File: .github/workflows/go.yml [hunk 4]
```diff
@@ -92,7 +94,7 @@ jobs:
   macos:
     name: AMD64 MacOS 10.15 Go ${{ matrix.go }}
     runs-on: macos-latest
-    if: github.event_name == 'push'
+    if: ${{ !contains(github.event.pull_request.title, '[skip ci]') }}
```
Comment (kou) [inline .github/workflows/go.yml]: Do we need this change?

**Review by kou — COMMENTED** [review]
Thanks for your summary!

I considered.
How about checking "WIP:" in pull request title or the draft state of pull request instead of "[skip ci]" in pull request title?
We may forget the "[skip ci]" in pull request title rule because there is no convention for this style. We may also forget remove "[skip ci]" from pull request title on merge.
But we will not forget the "WIP:" in pull request title rule because this is a widely used convention. We will not forget to remove "WIP:" from pull request title on merge. In fact, we used "WIP:" but we don't have merged commit that has "WIP" message recently:

```console
$ git log --oneline | grep WIP
492b3d439 ARROW-613: WIP TypeScript Implementation
16e44e3d4 ARROW-3: This patch includes a WIP draft specification document for the physical Arrow memory layout produced over a series of discussions amongst the to-be Arrow committers during late 2015. There are also a few small PNG diagrams that illustrate some of the Arrow layout concepts.
```

File: .github/workflows/csharp.yml (hunk 3 above)
Comment (kszucs) [reply .github/workflows/csharp.yml]: We don't have many C# patches so this won't fill our queue.

File: .github/workflows/go.yml (hunk 4 above)
Comment (kszucs) [reply .github/workflows/go.yml]: Same as with C#

**Comment by kszucs** [comment]
> Thanks for your summary!
> 
> I considered.
> How about checking "WIP:" in pull request title or the draft state of pull request instead of "[skip ci]" in pull request title?
Sounds good to me, although Appveyor won't skip the builds then.

File: .github/workflows/cpp_cron.yml (hunk 2 above)
Comment (kszucs) [reply .github/workflows/cpp_cron.yml]: It is already in cron, so we don't need to filter on the event.

**Comment by kou** [comment]
I think that we don't need to care about AppVeyor because:

  * AppVeyor has few jobs now
  * AppVeyor cancels pending jobs when we push new commits to pull request

**Comment by kszucs** [comment]
Updating the PR to use the `[WIP]` pattern instead.

**Comment by kszucs** [comment]
Ok, seems like no build has been triggered. Let's try out without the WIP pattern in the title.

**Review by pitrou — APPROVED** [review]
LGTM

File: ci/scripts/util_checkout.sh [hunk 5]
```diff
@@ -17,10 +17,9 @@
 # specific language governing permissions and limitations
 # under the License.
 
-# this script is github actions specific
+# this script is github actions specific to check out the submodules and tags
 
-# checkout the submodules, remove it once the "submodules: recursive" feature
-# gets released again
+# TODO(kszucs): remove it once the "submodules: recursive" feature is released
```
Comment (kou) [inline ci/scripts/util_checkout.sh:22]: https://github.com/actions/checkout/tree/v2
It's released 6 days ago. :-)

**Pull request closed without merging**